src/cfnlint/data/AdditionalSpecs/RequiredBasedOnValue.json

```diff
@@ -0,0 +1,22 @@
+{
+    "PropertyTypes": {},
+    "ResourceTypes": {
+        "AWS::Lambda::EventSourceMapping": {
+            "EventSourceArn": [
+                {
+                    "Regex": "arn:aws:(kinesis|kafka|dynamodb):*",
+                    "GetAtt": {
+                        "AWS::DynamoDB::Table": "StreamArn",
+                        "AWS::Kinesis::Stream": "Arn"
+                    },
+                    "Ref": [
+                        "AWS::MSK::Cluster"
+                    ],
+                    "RequiredProperties": [
+                        "StartingPosition"
+                    ]
+                }
+            ]
+        }
+    }
+}
```

Comment:
Like reading directly from [`AdditionalSpecs`](https://github.com/aws-cloudformation/cfn-python-lint/blob/master/src/cfnlint/data/AdditionalSpecs) and avoiding the patching process of [`ExtendedSpecs`](https://github.com/aws-cloudformation/cfn-python-lint/blame/master/src/cfnlint/data/ExtendedSpecs/) into [`CloudSpecs`](https://github.com/aws-cloudformation/cfn-python-lint/tree/master/src/cfnlint/data/CloudSpecs) and the ability to provide patterns in addition to literal strings 👍

Dislike JSON, separate `ResourceTypes`/`PropertyTypes` sections (should be inferable from type name alone and has caused confusion in the past: https://github.com/aws-cloudformation/cfn-python-lint/pull/1608/, https://github.com/aws-cloudformation/cfn-python-lint/pull/1654, https://github.com/aws-cloudformation/cfn-python-lint/pull/1712), and mapping [`Ref`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference-ref.html)/[`GetAtt`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference-getatt.html) per constraint

We should optimize for [concise constraints](https://github.com/aws-cloudformation/cfn-python-lint/issues/903#issuecomment-595356853) at the expense of potentially more one-time parsing code

```bash
# just an example possibility for discussion
StartingPosition if AWS::Lambda::EventSourceMapping.EventSourceArn == arn:aws*:(kinesis|kafka|dynamodb):*
```

Reply:
I like the readability of what you provided but yea we will need to think about that and how to parse it.  Let me try a few things to see if I can make that work well.

Reply:
Or maybe even just [schematize](https://json-schema.org/) some of these constraint files so we automatically catch mistakes

Reply:
I agree.  I removed the resource types and property types.  Let me schematize it and add testing for that file.

Reply:
Created a JSON Schema for the file and added testing.